sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java

```diff
@@ -36,29 +38,20 @@
 
     @Override
     public Mono<AccessToken> getToken(TokenRequestContext request) {
-        List<CredentialUnavailableException> exceptions = new ArrayList<>(4);
+        StringBuilder message = new StringBuilder();
         return Flux.fromIterable(credentials)
-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {
-                       exceptions.add(t);
+                   .flatMap(p -> p.getToken(request).onErrorResume(Exception.class, t -> {
+                       if (!t.getClass().getSimpleName().equals("CredentialUnavailableException")) {
+                           throw logger.logExceptionAsError(new CredentialUnavailableException(
+                                   unavailableError + p.getClass().getSimpleName() + " authentication failed.",
+                                   t));
+                       }
+                       message.append(t.getMessage()).append(" "); 
```

Comment:
A single large message will be unfriendly to read.
Better to keep the individual exceptions and their stack trace separate

Reply:
In this pr, I change java error message to java(new), and other languages' error message is now as follow, Do java need change it to java(new)? If you think java need change and the java(new) error message is not good enough, please  change it in the under table , and I will implement.
 
Scenario | Python/C# | Java(Current) | Java(new)
-- | -- | -- | --
Credential   Unavailable | DefaultAzureCredential   failed to retrieve a token from the included credentials.     EnvironmentCredential authentication unavailable. Environment variables are   not fully configured.     ManagedIdentityCredential authentication unavailable. No Managed Identity   endpoint found.     SharedTokenCacheCredential authentication unavailable. No accounts were   found in the cache. | Tried   EnvironmentCredential, ManagedIdentityCredential, SharedTokenCacheCredential,   AzureCliCredential but failed to acquire a token for any of them. Please   verify the environment for the credentials and see more details in the causes   below. | DefaultAzureCredential failed to retrieve a token from the   included credentials.EnvironmentCredential   authentication unavailable. Environment variables are not fully configured.   ManagedIdentityCredential authentication unavailable. Connection to IMDS   endpoint cannot be established, connect timed out. SharedTokenCacheCredential   authentication unavailable. No accounts were found in the cache.   AzureCliCredential authentication unavailable.
Authenticate   Failed | DefaultAzureCredential   authentication failed. ---> Azure.Identity.AuthenticationFailedException:   ClientSecretCredential authentication failed. --->   Azure.RequestFailedException: Service request failed.     Status: 400 (Bad Request)   Content:       {"error":"unauthorized_client","error_description":"AADSTS700016:   Application with identifier '78e8f606-0950-4210-906d-1d321511c09d' was not   found in the directory 'a7fc734e-9961-43ce-b4de-21b8b38403ba'. This can   happen if the application has not been installed by the administrator of the   tenant or consented to by any user in the tenant. You may have sent your   authentication request to the wrong tenant.\r\nTrace ID:   9f3624a7-49ca-4097-a339-780fd4e41600\r\nCorrelation ID:   b1acfa40-06cc-42a1-a2f1-d35b94db3628\r\nTimestamp: 2020-02-06   01:13:33Z","error_codes":[700016],"timestamp":"2020-02-06   01:13:33Z","trace_id":"9f3624a7-49ca-4097-a339-780fd4e41600","correlation_id":"b1acfa40-06cc-42a1-a2f1-d35b94db3628","error_uri":"https://login.microsoftonline.com/error?code=700016"} | Tried   EnvironmentCredential, ManagedIdentityCredential, SharedTokenCacheCredential,   AzureCliCredential but failed to acquire a token for any of them. Please   verify the environment for the credentials and see more details in the causes   below. | DefaultAzureCredential   authentication failed. ---> EnvironmentCredential authentication failed.   …….    Suppressed:   com.microsoft.aad.msal4j.MsalServiceException: AADSTS90002: Tenant   '72f988bf-86f1-41af-91ab-2d7cd011db41' not found. This may happen if there   are no active subscriptions for the tenant. Check to make sure you have the   correct tenant ID. Check with your subscription administrator.

Reply:
In this pr, I change java error message to java(new), and other languages' error message is now as follow, Do java need change it to java(new)? If you think java need change and the java(new) error message is not good enough, please  change it in the under table , and I will implement.
 
Scenario | Python/C# | Java(Current) | Java(new)
-- | -- | -- | --
Credential   Unavailable | DefaultAzureCredential   failed to retrieve a token from the included credentials.     EnvironmentCredential authentication unavailable. Environment variables are   not fully configured.     ManagedIdentityCredential authentication unavailable. No Managed Identity   endpoint found.     SharedTokenCacheCredential authentication unavailable. No accounts were   found in the cache. | Tried   EnvironmentCredential, ManagedIdentityCredential, SharedTokenCacheCredential,   AzureCliCredential but failed to acquire a token for any of them. Please   verify the environment for the credentials and see more details in the causes   below. | DefaultAzureCredential failed to retrieve a token from the   included credentials.EnvironmentCredential   authentication unavailable. Environment variables are not fully configured.   ManagedIdentityCredential authentication unavailable. Connection to IMDS   endpoint cannot be established, connect timed out. SharedTokenCacheCredential   authentication unavailable. No accounts were found in the cache.   AzureCliCredential authentication unavailable.
Authenticate   Failed | DefaultAzureCredential   authentication failed. ---> Azure.Identity.AuthenticationFailedException:   ClientSecretCredential authentication failed. --->   Azure.RequestFailedException: Service request failed.     Status: 400 (Bad Request)   Content:       {"error":"unauthorized_client","error_description":"AADSTS700016:   Application with identifier '78e8f606-0950-4210-906d-1d321511c09d' was not   found in the directory 'a7fc734e-9961-43ce-b4de-21b8b38403ba'. This can   happen if the application has not been installed by the administrator of the   tenant or consented to by any user in the tenant. You may have sent your   authentication request to the wrong tenant.\r\nTrace ID:   9f3624a7-49ca-4097-a339-780fd4e41600\r\nCorrelation ID:   b1acfa40-06cc-42a1-a2f1-d35b94db3628\r\nTimestamp: 2020-02-06   01:13:33Z","error_codes":[700016],"timestamp":"2020-02-06   01:13:33Z","trace_id":"9f3624a7-49ca-4097-a339-780fd4e41600","correlation_id":"b1acfa40-06cc-42a1-a2f1-d35b94db3628","error_uri":"https://login.microsoftonline.com/error?code=700016"} | Tried   EnvironmentCredential, ManagedIdentityCredential, SharedTokenCacheCredential,   AzureCliCredential but failed to acquire a token for any of them. Please   verify the environment for the credentials and see more details in the causes   below. | DefaultAzureCredential   authentication failed. ---> EnvironmentCredential authentication failed.   …….    Suppressed:   com.microsoft.aad.msal4j.MsalServiceException: AADSTS90002: Tenant   '72f988bf-86f1-41af-91ab-2d7cd011db41' not found. This may happen if there   are no active subscriptions for the tenant. Check to make sure you have the   correct tenant ID. Check with your subscription administrator.

Reply:
We need to retain the stack traces from individual exception and chain them.
Java devs use the stack trace to analyze / debug issues.